### ISSUE TITLE
Parent thread is never GC'd after its termination

### ISSUE THREAD

**Issue by b-inary** [GoogleChromeLabs/wasm-bindgen-rayon#37]
I suppose this is a difficult issue to address, but I am posting this for those having the same problem.

As the title says, the parent worker thread is never GC'd by the browser after we terminate it or even after we reload the page. Perhaps this is a regression in Google Chrome 109, as I don't think I have experienced this before. Firefox 108 may or may not be able to GC, depending on some conditions.

This can be confirmed by the following steps:

1. Open the wasm-bindgen-rayon demo page in Google Chrome 109.
2. Open the DevTools and select the Sources tab.
3. Reload the page.
4. Close the DevTools and reopen the DevTools.
5. You can see multiple "419.index.js".

To resolve this memory leak, I had to do the following workaround:

1. Since we cannot drop the global rayon thread pool, avoid using `build_global()` and create a local thread pool instead. We call multithreaded functions by `THREAD_POOL.as_ref().map(|pool| pool.install(|| do_something()));`.
```rust
pub static mut THREAD_POOL: Option<rayon::ThreadPool> = None;

#[wasm_bindgen]
impl wbg_rayon_PoolBuilder {
    ...
    pub fn build(&mut self) {
        unsafe {
            THREAD_POOL = Some(...);
        }
    }
}
```
2. Modify the `workerHelpers.js`.
```diff
 waitForMsgType(self, 'wasm_bindgen_worker_init').then(async data => {
   ...
   pkg.wbg_rayon_start_worker(data.receiver);
+  postMessage({ type: 'wasm_bindgen_worker_done' });
 }
```
3. Before terminating the parent worker, call the following `exitThreadPool` function.
```rust
#[wasm_bindgen(js_name = exitThreadPool)]
pub fn exit_thread_pool() -> Promise {
    unsafe {
        let promise = terminate_workers();
        THREAD_POOL = None;
        promise
    }
}
```
```js
export function terminateWorkers() {
  return Promise.all(
    _workers.map(async (worker) => {
      await waitForMsgType(worker, 'wasm_bindgen_worker_done');
      worker.terminate();
    })
  );
}
```
4. Terminate the parent worker after some delay.
```js
setTimeout(() => parentWorker.terminate(), 1000);
```

The workaround described above allows workers to terminate without causing memory leaks on both Chrome and Firefox. However, I do not know what to do about page reloading, as setting listeners on the `beforeunload` or `pagehide` events did not help in my environment.

**Comment by RReverser** [comment]
Interesting, @jakearchibald just posted recently about seeing similar issue in Safari TP, but I also don't think we saw it in Chrome before: https://twitter.com/jaffathecake/status/1613901682437505025

How do you notice / measure the leak?

**Comment by RReverser** [comment]
> 5\. You can see multiple "419.index.js".

FWIW I can reproduce this, but it's weird that the number of child workers stays the same - you'd think those are the ones that wouldn't be GC'd. It's also odd that DevTools shows only one 419.index.js until closed & reopened.

**Comment by b-inary** [comment]
I think there is an actual memory leak. I am publishing [a web page](https://github.com/b-inary/wasm-postflop) that performs calculations with multithreaded WASM. I have recently noticed an increase in memory usage if I solve another instance after terminating the previous one.

> FWIW I can reproduce this, but it's weird that the number of child workers stays the same - you'd think those are the ones that wouldn't be GC'd. It's also odd that DevTools shows only one 419.index.js until closed & reopened.

In my environment, children workers are GC'd correctly. I agree that there is something wrong with the browser side.

**Comment by RReverser** [comment]
Hm and yeah I can't reproduce this with Chromium 108.

**Comment by RReverser** [comment]
I don't think I can do much from my side, can you please report it to crbug.com with repro steps? (no need to mention workaround I think, just link back to the demo & this issue)

**Comment by RReverser** [comment]
It's also a bit weird that, if the DevTools is closed and I try to reload the page several times, and then open DevTools, I don't see multiple 419.index.js, so it still seems likely that it's only a DevTools-related bug.

**Comment by RReverser** [comment]
Just tried @jakearchibald's demo for Safari repro: https://safari-worker-memory-leak-demo.netlify.app/

In his variation he recreates the threadpool w/o reloading the page (on each click), and I can very clearly see the top worker remaining alive in Chrome 109 but not Chrome 108.

**Comment by RReverser** [comment]
Seems fixed in Chrome Canary 111.0.5540.0 though, so maybe there's something that needs to be backported.

**Comment by b-inary** [comment]
Thank you so much for your quick response and intensive investigation.

> Seems fixed in Chrome Canary 111.0.5540.0

Then, should I report this to crbug? I will do my best if necessary, but as I have not reported a bug before, I do not know how to go about it. I apologize for bothering you.

**Comment by RReverser** [comment]
> Then, should I report this to crbug?

Not sure either tbh. If it's affecting production, I suppose there's no harm in reporting to see if they'd backport the fix.

Either way, closing on this side.
